src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs

```diff
@@ -25,20 +25,24 @@ public static byte[] CloneByteArray(this byte[] src)
         public static char[] ToHexArrayUpper(this byte[] bytes)
         {
             char[] chars = new char[bytes.Length * 2];
+            ToHexArrayUpper(bytes, chars);
+            return chars;
+        }
+
+        private static void ToHexArrayUpper(byte[] bytes, Span<char> chars)
+        {
+            Debug.Assert(chars.Length >= bytes.Length * 2);
             int i = 0;
             foreach (byte b in bytes)
             {
                 chars[i++] = NibbleToHex((byte)(b >> 4));
                 chars[i++] = NibbleToHex((byte)(b & 0xF));
             }
-            return chars;
         }
 
         // Encode a byte array as an upper case hex string.
-        public static string ToHexStringUpper(this byte[] bytes)
-        {
-            return new string(ToHexArrayUpper(bytes));
-        }
+        public static string ToHexStringUpper(this byte[] bytes) =>
+            string.Create(bytes.Length * 2, bytes, (chars, src) => ToHexArrayUpper(src, chars));
```

Comment:
Would it make more sense to flip the parameters of the private so the lambda isn't applying a twist? (I don't know if the JIT is smart enough to rewrite the parameter order of the private to fix that)

Reply:
Ok.

Reply:
I just tried, and it looks very strange to me to have the destination first, especially when they're named "chars" and "bytes" and so it's not clear which is the src and which is the dst.  I can rename them, but... I think I'll just leave it if that's ok with you.

Reply:
Yeah, I'm fine either way. It just popped in as "maybe this would help the inliner", because I happened to be thinking about how weird inlining algorithms must be earlier today.